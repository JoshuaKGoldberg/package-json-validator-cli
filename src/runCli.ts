import * as fsSync from "node:fs";
import * as fs from "node:fs/promises";
import { validate } from "package-json-validator";
import yargs from "yargs";

interface Options {
	filename: string;
	quiet: boolean;
	recommendations: boolean;
	warnings: boolean;
}

export async function runCli(args: string[]) {
	const options = yargs(args)
		.options("filename", {
			alias: "f",
			default: "package.json",
			description: "package.json file to validate",
		})
		.options("warnings", {
			alias: "w",
			boolean: true,
			default: false,
			description: "display warnings",
		})
		.options("recommendations", {
			alias: "r",
			boolean: true,
			default: false,
			description: "display recommendations",
		})
		.options("quiet", {
			alias: "q",
			boolean: true,
			default: false,
			description: "less output",
		})
		.usage("Validate package.json files")
		.parse() as Options;

	if (!fsSync.existsSync(options.filename)) {
		console.error("File does not exist: " + options.filename);
		return 1;
	}

	const contents = (await fs.readFile(options.filename)).toString();
	const results = validate(contents, {
		recommendations: options.recommendations,
		warnings: options.warnings,
	});

	if (!results.valid) {
		console.error(options.filename + " is NOT valid");
		console.error(results);
		return 1;
	}

	if (!options.quiet) {
		console.log(results);
	}

	return 0;
}
