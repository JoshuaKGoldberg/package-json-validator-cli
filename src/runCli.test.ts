import { beforeEach, describe, expect, it, MockInstance, vi } from "vitest";

import { runCli } from "./runCli.ts";

const mockExistsSync = vi.fn();
const mockReadFile = vi.fn();

vi.mock("node:fs", () => ({
	get existsSync() {
		return mockExistsSync;
	},
}));

vi.mock("node:fs/promises", () => ({
	get readFile() {
		return mockReadFile;
	},
}));

let mockError: MockInstance;
let mockLog: MockInstance;

beforeEach(() => {
	mockError = vi.spyOn(console, "error").mockImplementation(() => undefined);
	mockLog = vi.spyOn(console, "log").mockImplementation(() => undefined);
});

const contentsInvalid = { version: 0 };
const contentsValid = { name: "example", version: "0.0.0" };

describe(runCli, () => {
	it("outputs an error and returns 1 when package.json does not exist and --filename is not provided", async () => {
		mockExistsSync.mockReturnValueOnce(false);

		const actual = await runCli([]);

		expect(actual).toBe(1);
		expect(mockError).toHaveBeenCalledWith("File does not exist: package.json");
		expect(mockExistsSync).toHaveBeenCalledWith("package.json");
	});

	it("outputs an error and returns 1 when the --filename package does not exist", async () => {
		mockExistsSync.mockReturnValueOnce(false);

		const mockError = vi
			.spyOn(console, "error")
			.mockImplementation(() => undefined);

		const actual = await runCli(["--filename", "other.json"]);

		expect(actual).toBe(1);
		expect(mockError).toHaveBeenCalledWith("File does not exist: other.json");
		expect(mockExistsSync).toHaveBeenCalledWith("other.json");
	});

	it("outputs the results and returns 1 when the default package.json is not valid", async () => {
		const contents = JSON.stringify(contentsInvalid);

		mockExistsSync.mockReturnValueOnce(true);
		mockReadFile.mockReturnValue(contents);

		const mockError = vi
			.spyOn(console, "error")
			.mockImplementation(() => undefined);

		const actual = await runCli([]);

		expect(actual).toBe(1);
		expect(mockError.mock.calls).toMatchSnapshot();
		expect(mockReadFile).toHaveBeenCalledWith("package.json");
	});

	it("outputs the results and returns 1 when a --filename package.json is not valid", async () => {
		const contents = JSON.stringify(contentsInvalid);

		mockExistsSync.mockReturnValueOnce(true);
		mockReadFile.mockReturnValue(contents);

		const mockError = vi
			.spyOn(console, "error")
			.mockImplementation(() => undefined);

		const actual = await runCli(["--filename", "other.json"]);

		expect(actual).toBe(1);
		expect(mockError.mock.calls).toMatchSnapshot();
		expect(mockReadFile).toHaveBeenCalledWith("other.json");
	});

	it("returns 0 and logs results when the file is valid and --quiet is not provided", async () => {
		const contents = JSON.stringify(contentsValid);

		mockExistsSync.mockReturnValueOnce(true);
		mockReadFile.mockReturnValue(contents);

		const mockError = vi
			.spyOn(console, "error")
			.mockImplementation(() => undefined);

		const actual = await runCli([]);

		expect(actual).toBe(0);
		expect(mockError).not.toHaveBeenCalled();
		expect(mockLog.mock.calls).toMatchSnapshot();
	});

	it("returns 0 without logging results when the file is valid and --quiet is provided", async () => {
		const contents = JSON.stringify(contentsValid);

		mockExistsSync.mockReturnValueOnce(true);
		mockReadFile.mockReturnValue(contents);

		const mockError = vi
			.spyOn(console, "error")
			.mockImplementation(() => undefined);

		const actual = await runCli(["--quiet"]);

		expect(actual).toBe(0);
		expect(mockError).not.toHaveBeenCalled();
		expect(mockLog).not.toHaveBeenCalled();
	});
});
