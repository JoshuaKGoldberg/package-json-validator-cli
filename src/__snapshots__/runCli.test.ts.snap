// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runCli > outputs the results and returns 1 when a --filename package.json is not valid 1`] = `
[
  [
    "other.json is NOT valid",
  ],
  [
    {
      "errors": [
        {
          "field": "name",
          "message": "Missing required field: name",
        },
        {
          "field": "version",
          "message": "the type should be a \`string\`, not \`number\`",
        },
      ],
      "valid": false,
    },
  ],
]
`;

exports[`runCli > outputs the results and returns 1 when the default package.json is not valid 1`] = `
[
  [
    "package.json is NOT valid",
  ],
  [
    {
      "errors": [
        {
          "field": "name",
          "message": "Missing required field: name",
        },
        {
          "field": "version",
          "message": "the type should be a \`string\`, not \`number\`",
        },
      ],
      "valid": false,
    },
  ],
]
`;

exports[`runCli > returns 0 and logs results when the file is valid and --quiet is not provided 1`] = `
[
  [
    {
      "valid": true,
    },
  ],
]
`;
